tls/s2n_server_hello.c

```diff
@@ -36,6 +37,54 @@
 /* From RFC5246 7.4.1.2. */
 #define S2N_TLS_COMPRESSION_METHOD_NULL 0
 
+/* From RFC8446 4.1.3. */
+#define S2N_DOWNGRADE_PROTECTION_SIZE   8
+const uint8_t tls12_downgrade_protection_bytes[] = {
+    0x44, 0x4F, 0x57, 0x4E, 0x47, 0x52, 0x44, 0x01
+};
+
+const uint8_t tls11_downgrade_protection_bytes[] = {
+    0x44, 0x4F, 0x57, 0x4E, 0x47, 0x52, 0x44, 0x00
+};
+
+static int s2n_detect_downgrade(struct s2n_connection *conn) {
+    notnull_check(conn);
+    uint8_t *downgrade_protection = &conn->secure.server_random[S2N_TLS_RANDOM_DATA_LEN - S2N_DOWNGRADE_PROTECTION_SIZE];
+
+    /* Detect downgrade attacks according to RFC 8446 section 4.1.3 */
+    if (s2n_is_tls13_enabled() && conn->server_protocol_version == S2N_TLS12) {
```

Comment:
i believe client_protocol_version should also be checked

Reply:
Looking at this again, this check just generally looks wrong. The RFC says, "If negotiating TLS 1.1 or below, TLS 1.3 servers MUST", but we don't know that we're talking to a tls1.3 server here. We just know that we're a tls 1.3 client, and that the server wants to talk 1.2. The only way we should alert is if we receive a supported_versions extension which lists tls 1.2 as the max supported version-- then we know we're talking to a tls1.3 server.

Reply:
I don't explicitly check the client protocol because this is run in the client context. So if tls13_enabled() is true, then we should have negotiated tls13. Otherwise we should have negotiated tls12. If we negotiated lower than what the client can support, we have to check for the downgrade mechanism.

Reply:
@lrstewart We know we're talking to a downgraded server if the random bytes are set as indicated in the RFC. If the random data is truly random, then we are talking to a tls12 server or a non-compliant tls13 server.

Reply:
Okay, ignore that comment. I reread the RFC. We need to alert when:
1. The client supports TLS1.3.
2. The server negotiates TLS1.2
3. The server sets the downgrade string, indicating that it could have negotiated TLS1.3 if given the option
4. The client KNOWS it sent TLS1.3 as an option.

If we can assume we always send TLS1.3 as an option when s2n_is_tls13_enabled() is true, then we are good as-is. But I'm not sure that's always true.

Reply:
s2n_is_tls13_enabled() currently means the lib support up to tls13, so we have to be careful with not to misuse "s2n_is_tls13_enabled()" with "is this a tls13 connection".

Reply:
I think I've got the differences correct now. Updated in the next push.